README.md
## web3_provider
The project supported send and receive messages between Dapp and in-app webview “Only EIP-1193 standard supported”

#### Plugin: [https://pub.dev/packages/web3_provider](https://pub.dev/packages/web3_provider)

## Overview

* Communicate between dapp and your app.

<img src="https://raw.githubusercontent.com/VNAPNIC/web3-provider/master/art/sequence_diagram_communication.png"/>

* Dapp interact with blockchain.

<img src="https://raw.githubusercontent.com/VNAPNIC/web3-provider/master/art/sequence_diagram_sign_transaction.png"/>

## Usage

```dart
import 'package:web3_provider/web3_provider.dart';

InAppWebViewGroupOptions options = InAppWebViewGroupOptions(
  crossPlatform: InAppWebViewOptions(
    useShouldOverrideUrlLoading: true,
    mediaPlaybackRequiresUserGesture: false,
    userAgent:
    "Mozilla/5.0 (Linux; Android 4.4.4; SAMSUNG-SM-N900A Build/tt) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/33.0.0.0 Mobile Safari/537.36",
  ),
  android: AndroidInAppWebViewOptions(
    useHybridComposition: true,
    domStorageEnabled: true,
  ),
  ios: IOSInAppWebViewOptions(
    allowsInlineMediaPlayback: true,
  ),
);

/// By default config
InAppWebViewEIP1193(
    chainId: 56, // Replace your chain id network you want connect
    rpcUrl: 'https://bsc-dataseed.binance.org/', // Replace your rpc url network you want connect
    walletAddress: walletAddress,
    signCallback: (rawData, eip1193, controller) {
      // Handler callback when dapp interact with blockchain
      switch (eip1193) {
        case EIP1193.requestAccounts:
          controller?.setAddress(walletAddress, id);
          print('requestAccounts');
          break;
        case EIP1193.signTransaction:
          print('signTransaction');
          break;
        case EIP1193.signMessage:
          print('signMessage');
          break;
        case EIP1193.signPersonalMessage:
          print('signPersonalMessage');
          break;
        case EIP1193.signTypedMessage:
          print('addEthereumChain');
          break;
        case EIP1193.addEthereumChain:
          print('addEthereumChain');
          break;  
    },
    initialUrlRequest: URLRequest(
        url: Uri.parse(
        'https://position.exchange', // Replace your dapp domain
        ),
    ),
    initialOptions: options
);
```

If you want use your provider script
you provide [customPathProvider] and [customWalletName]

`signCallback: (rawData, eip1193, controller)`: callback was called when dapp when interact with blockchain. <br/>
- `rawData`: data received.
- `eip1193`: type function support.
  - requestAccounts: Pass when web app connect wallet
  - signTransaction: Pass when web app approve contract or send transaction
  - signMessage: Pass when web app sign a message
  - signPersonalMessage: Pass when web app sign a personal message
  - signTypedMessage: Pass when web app sign a type message
  - addEthereumChain: Pass when web app add a new chain


When you pass data from dapp to your app
```
const args = {/* Pass data you want */};
if (window.flutter_inappwebview.callHandler) {
    window.flutter_inappwebview.callHandler('functionName', args)
        .then(function(result) {
          /// Receive data from your app
    });
}
```

And receive in your app
```
onWebViewCreated: (controller) {
    controller.addJavaScriptHandler(
        handlerName: 'functionName',
        callback: (args) {
          /// Receive data from dapp
          
          
          /// Send data to dapp;
          return /* anything */;
        },
    );
},
```


CHANGELOG.md
## 1.1.2
* update mechanism handle request via EIP1193

## 1.1.1
* update version flutter_inappwebview

## 1.1.0
* allow custom provider

## 1.0.3
* connect to new wallet 

## 1.0.2

* update docs

## 1.0.1

* update docs
* update path provider

## 1.0.0

* init plugin


lib/json_util.dart
import 'dart:convert';

import 'package:flutter/foundation.dart';

/// Json Util.
class JsonUtil {
  /// Converts object [value] to a JSON string.
  static String? encodeObj(Object value) {
    return json.encode(value);
  }

  /// Converts JSON string [source] to object.
  static dynamic getObj<T>(String? source) {
    if (source == null || source.isEmpty) return null;
    try {
      return json.decode(source);
    } catch (e) {
      return null;
    }
  }

  static Future<dynamic> getObjFuture<T>(String? source) async {
    if (source == null || source.isEmpty) return null;
    try {
      return await compute(getObj, source);
    } catch (e) {
      return null;
    }
  }

  /// Converts JSON string list [source] to object list.
  static List<T>? getObjList<T>(String source) {
    if (source.isEmpty) return null;
    try {
      List? list = json.decode(source);

      return list?.map((value) {
        return value;
      }).toList() as List<T>?;
    } catch (e) {
      print('JsonUtil convert error, Exception：${e.toString()}');
    }

    return null;
  }
}


lib/formatter_balance.dart
import 'package:decimal/decimal.dart';

/// Extension BigInt
extension BigIntExt on BigInt {
  /// Format bigint to string with [decimals]
  String tokenString(int decimals) {
    Decimal value = Decimal.fromBigInt(this);
    Decimal decimalValue =
        (value / Decimal.fromInt(10).pow(decimals).toDecimal()).toDecimal();
    return decimalValue.toString();
  }
}

class FormatterBalance {
  /// Calculate gas price to send transaction
  static String configFeeValue({
    required String? beanValue,
    required String? offsetValue,
  }) {
    beanValue = beanValue?.replaceAll("null", "0");
    offsetValue = offsetValue?.replaceAll("null", "0");
    Decimal gasValue = Decimal.parse(beanValue ?? "0") *
        Decimal.fromInt(10).pow(9).toDecimal();
    gasValue = gasValue * Decimal.parse(offsetValue ?? "0");
    var feeValue =
        (gasValue / Decimal.fromInt(10).pow(18).toDecimal()).toDecimal();
    return feeValue.toStringAsFixed(6);
  }
}


lib/js_bridge_callback_bean.dart
import 'package:flutter_inappwebview/flutter_inappwebview.dart';

/// Extension's [InAppWebViewController]
extension Web3Result on InAppWebViewController {
  /// send [Canceled] error to dapp
  void cancel(int id) {
    sendError("Canceled", id);
  }

  /// send error to dapp
  void sendError(String error, int id) {
    final script = "window.ethereum.sendError($id,\"$error\")";
    evaluateJavascript(source: script);
  }

  /// send result to dapp
  void sendResult(String result, int id) {
    final script = "window.ethereum.sendResponse($id, \"$result\")";
    evaluateJavascript(source: script);
  }

  /// send string list result to dapp
  void sendResults(List<String> results, int id) {
    final array = results
        .map((e) => e.toLowerCase())
        .toList()
        .map((e) => "\"$e\"")
        .toList();
    final arrayStr = array.join(",");
    final script = "window.ethereum.sendResponse($id, [$arrayStr])";
    evaluateJavascript(source: script);
  }

  /// set address connect to dapp
  void setAddress(String address, int id) async {
    address = address.toLowerCase();
    final script = "window.ethereum.setAddress('$address');";
    await evaluateJavascript(source: script);
    sendResults([address], id);
  }
}


example/lib/communicate_dapp.dart
import 'dart:convert';

import 'package:flutter/material.dart';
import 'package:web3_provider/web3_provider.dart';

class CommunicateDapp extends StatefulWidget {
  const CommunicateDapp({Key? key}) : super(key: key);

  @override
  State<CommunicateDapp> createState() => _CommunicateDappState();
}

class _CommunicateDappState extends State<CommunicateDapp> {
  InAppWebViewGroupOptions options = InAppWebViewGroupOptions(
    crossPlatform: InAppWebViewOptions(
      useShouldOverrideUrlLoading: true,
      mediaPlaybackRequiresUserGesture: false,
      userAgent:
          "Mozilla/5.0 (Linux; Android 4.4.4; SAMSUNG-SM-N900A Build/tt) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/33.0.0.0 Mobile Safari/537.36",
    ),
    android: AndroidInAppWebViewOptions(
      useHybridComposition: true,
      domStorageEnabled: true,
    ),
    ios: IOSInAppWebViewOptions(
      allowsInlineMediaPlayback: true,
    ),
  );

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: InAppWebView(
        initialData: InAppWebViewInitialData(
          data: """
                <!DOCTYPE html>
                <html lang="en">
                    <head>
                        <meta charset="UTF-8">
                        <meta name="viewport" content="width=device-width, user-scalable=no, initial-scale=1.0, maximum-scale=1.0, minimum-scale=1.0">
                    </head>
                    <body>
                        <button onclick="clickFunction()">Click me</button>
                        <script>
                            function clickFunction() {
                                const args = 
                                {
                                  param : {
                                    url: "https://rpc.ankr.com/bsc/206242aedbcf68f4c731e89f2988a434fb7e46560bee0ba90e4a2370b5ec71a9", 
                                    method: "POST", 
                                    headers: {

                                    }, 
                                    body: {
                                      "jsonrpc": "2.0",
                                      "id": 800061,
                                      "method": "eth_call",
                                      "params": [
                                        {
                                          "data": "0x252dba420000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000200000000000000000000000005ca42204cdaa70d5c773946e69de942b85ca67060000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000002470a0823100000000000000000000000062a9b1ab58c3b59ef17923792122985d210a94a100000000000000000000000000000000000000000000000000000000",
                                          "to": "0x1ee38d535d541c55c9dae27b12edf090c608e6fb"
                                        },
                                        "latest"
                                      ]
                                    }
                                  }
                                };
                                if (window.flutter_inappwebview.callHandler) {
                                  window.flutter_inappwebview.callHandler('clickFunction', args)
                                    .then(function(result) {
                                        alert(result);
                                    });
                                }
                            }
                        </script>
                    </body>
                </html>                              
                """,
        ),
        initialOptions: options,
        onWebViewCreated: (controller) {
          controller.addJavaScriptHandler(
            handlerName: 'clickFunction',
            callback: (args) {
              Map<String, dynamic> param = args[0]['param'];
              String method = param['method'];
              String url = param['url'];
              Map<String, dynamic>? headers = param['headers'];
              Map<String, dynamic>? body = param['body'];

              /// Handle with data received.

              /// Pass to dapp
              final result = {
                'url': url,
                'headers': headers,
                'body': body,
              };

              return jsonEncode(result);
            },
          );
        },
        onConsoleMessage: (controller, consoleMessage) {
          print(consoleMessage);
        },
      ),
    );
  }
}


example/lib/main.dart
import 'dart:typed_data';

import 'package:eth_sig_util/util/utils.dart';
import 'package:flutter/material.dart';
import 'package:untitled/communicate_dapp.dart';
import 'package:web3_provider/web3_provider.dart';
import 'package:web3dart/json_rpc.dart';
import 'package:web3dart/web3dart.dart';
import 'package:http/http.dart';

import 'js_bridge_bean.dart';
import 'widget/payment_sheet_page.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  var apiUrl = "https://bsc-dataseed.binance.org/"; //Replace with your API
  var httpClient = Client();
  String walletAddress = '0x81D1aB023A46Fb4d2572533278b4131192cC5d45';
  final String prvKey =
      '0a8a8b091ae22bef656b3910b1541321a95fd62b6ea3ef4a8cc1589abd4f0d2a';
  late Web3Client ethClient;
  InAppWebViewController? _webViewController;

  InAppWebViewGroupOptions options = InAppWebViewGroupOptions(
    crossPlatform: InAppWebViewOptions(
      useShouldOverrideUrlLoading: true,
      mediaPlaybackRequiresUserGesture: false,
      userAgent:
          "Mozilla/5.0 (Linux; Android 4.4.4; SAMSUNG-SM-N900A Build/tt) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/33.0.0.0 Mobile Safari/537.36",
    ),
    android: AndroidInAppWebViewOptions(
      useHybridComposition: true,
      domStorageEnabled: true,
    ),
    ios: IOSInAppWebViewOptions(
      allowsInlineMediaPlayback: true,
    ),
  );

  _signTransaction({
    required BridgeParams bridge,
    required int chainId,
    required VoidCallback cancel,
    required Function(String idHash) success,
  }) async {
    final credentials = EthPrivateKey.fromHex(prvKey);
    final sender = EthereumAddress.fromHex(bridge.from ?? '');
    final signto = EthereumAddress.fromHex(bridge.to ?? '');
    final input = hexToBytes(bridge.data ?? '');
    String? price = (bridge.gasPrice == null)
        ? (await ethClient.getGasPrice()).toString()
        : bridge.gasPrice;
    int? maxGas;
    try {
      maxGas = (bridge.gas ??
          await ethClient.estimateGas(
            sender: sender,
            to: signto,
            data: input,
          )) as int?;
    } catch (e) {
      RPCError err = e as RPCError;
      cancel.call();
      return;
    }
    String fee = FormatterBalance.configFeeValue(
        beanValue: maxGas.toString(), offsetValue: price.toString());
    _showModalConfirm(
        from: walletAddress,
        to: bridge.to ?? '',
        value: bridge.value ?? BigInt.zero,
        fee: fee,
        confirm: () async {
          try {
            String result = await ethClient.sendTransaction(
              credentials,
              Transaction(
                  to: signto,
                  value: EtherAmount.inWei(bridge.value ?? BigInt.zero),
                  gasPrice: null,
                  maxGas: maxGas,
                  data: input),
              chainId: chainId,
              fetchChainIdFromNetworkId: false,
            );
            success.call(result);
          } catch (e) {
            if (e.toString().contains('-32000')) {
              ScaffoldMessenger.of(context).showSnackBar(const SnackBar(
                content: Text("gasLow"),
              ));
            } else {
              ScaffoldMessenger.of(context).showSnackBar(SnackBar(
                content: Text(e.toString()),
              ));
            }
          }
        },
        cancel: () {
          cancel.call();
        });
  }

  _showModalConfirm({
    required String from,
    required String to,
    required BigInt value,
    required String fee,
    required VoidCallback confirm,
    required VoidCallback cancel,
  }) {
    showModalBottomSheet(
        context: context,
        elevation: 0,
        isDismissible: true,
        isScrollControlled: true,
        shape: const RoundedRectangleBorder(
          borderRadius: BorderRadius.only(
            topLeft: Radius.circular(20),
            topRight: Radius.circular(20),
          ),
        ),
        builder: (_) {
          return PaymentSheet(
            datas: PaymentSheet.getTransStyleList(
              from: from,
              to: to,
              remark: '',
              fee: "$fee BNB",
            ),
            amount: "${value.tokenString(18)} BNB",
            nextAction: () async {
              confirm.call();
            },
            cancelAction: () {
              cancel.call();
            },
          );
        });
  }

  @override
  void initState() {
    super.initState();
    ethClient = Web3Client(apiUrl, httpClient);
  }

  @override
  Widget build(BuildContext context) {
    int chainId = 56;
    return Scaffold(
      appBar: AppBar(
        actions: [
          IconButton(
            onPressed: () {
              walletAddress = '0x8F0c3C6C466d2801F6D88002948a2d845f56B129';
              setState(() {});
              _webViewController?.reload();
            },
            icon: const Icon(Icons.add),
          )
        ],
        title: Text(widget.title),
      ),
      body: InAppWebViewEIP1193(
          customPathProvider: 'assets/example-trust-min.js',
          customWalletName: 'trustwallet',
          chainId: chainId,
          rpcUrl: 'https://bsc-dataseed.binance.org/',
          walletAddress: walletAddress,
          signCallback: (params, eip1193, controller) {
            final id = params["id"];
            switch (eip1193) {
              case EIP1193.requestAccounts:
                controller?.setAddress(walletAddress, id);
                print('requestAccounts');
                break;
              case EIP1193.signTransaction:
                Map<String, dynamic> object = params["object"];
                BridgeParams bridge = BridgeParams.fromJson(object);
                _signTransaction(
                    bridge: bridge,
                    chainId: chainId,
                    cancel: () {
                      controller?.cancel(id);
                    },
                    success: (idHash) {
                      controller?.sendResult(idHash, id);
                    });
                print('signTransaction');
                break;
              case EIP1193.signMessage:
              case EIP1193.signPersonalMessage:
                Map<String, dynamic> object = params["object"];
                String data = object["data"];
                _showModalConfirm(
                    from: walletAddress,
                    to: '',
                    value: BigInt.zero,
                    fee: '0',
                    confirm: () async {
                      final credentials = EthPrivateKey.fromHex(prvKey);
                      Uint8List message = await credentials
                          .signPersonalMessage(hexToBytes(data));
                      String result = bytesToHex(message, include0x: true);
                      controller?.sendResult(result, id);
                    },
                    cancel: () {
                      controller?.cancel(id);
                    });
                break;
              case EIP1193.signTypedMessage:
                Map<String, dynamic> object = params["object"];
                String raw = object["raw"];
                _showModalConfirm(
                    from: walletAddress,
                    to: '',
                    value: BigInt.zero,
                    fee: '0',
                    confirm: () async {
                      final credentials = EthPrivateKey.fromHex(prvKey);
                      Uint8List message =
                          await credentials.sign(hexToBytes(raw));
                      String result = bytesToHex(message, include0x: true);
                      controller?.sendResult(result, id);
                    },
                    cancel: () {
                      controller?.cancel(id);
                    });
                break;
              case EIP1193.addEthereumChain:
                print('addEthereumChain');
                break;
            }
          },
          initialUrlRequest: URLRequest(
            url: Uri.parse(
              'https://pancakeswap.finance/',
            ),
          ),
          initialOptions: options,
          onWebViewCreated: (controller) {
            _webViewController = controller;
          }),
    );
  }
}
